flax/linen/module.py

```diff
@@ -208,10 +208,10 @@ def wrapped_module_method(self, *args, **kwargs):
     is_compact_method = hasattr(fun, 'compact')
     is_setup_method = fun.__name__ == 'setup'
 
-    if self.scope is None:
```

Comment:
None of the scope helper methods `make_rng`, `variables`, `param` etc. do not check whether `scope is None`. It will now throw a confusing errror cannot access member of None.

Reply:
Instead of checking for `scope is None` everywhere, could we initialize `scope` with a dummy object that throws an exception with intelligible error message when any of its attributes is accessed?

Reply:
See my comment above, I think you either do or should get an error about being in a non-compact method.

Reply:
> Instead of checking for `scope is None` everywhere, could we initialize scope with a dummy object that throws an exception with intelligible error message when any of its attributes is accessed?

I like that. I'll first add tests and then consider this change once the test coverage is comprehensive.

Reply:
Actually after looking at the code more, there are many methods in Scope and implementing a new "Dummy object with good errors" there would be more code than just making sure we check `if scope is None` in a few more places. I'll be adding the relevant tests regardless so if we really wanted to later we could switch to this proposal later.

Reply:
What about subclassing `Scope` with an object that dynamically modifies these methods to rise the exception when they're called?

Reply:
Perhaps, but then it's harder to make the error messages explicit... They would have to be automatically generated from the method name, which may be fine, but then again we have some methods and `variables` is not a method but a property... Just sounds like a bit too much magic and indirection for something that can be solved with 3 more if statements and explicit errors

Reply:
Might be more code, but would make the `Module` code leaner and logically separate these checks into a proxy. And also protect against any such access in the future... Anyways, not feeling strongly about this, just throwing around an idea ;) Both solutions LGTM